homeassistant/components/hdmi_cec.py

```diff
@@ -4,118 +4,535 @@
-            _CEC.Transmit(_CEC.CommandFromString(cmd))
-        _LOGGER.info("Selected %s", call.data[ATTR_DEVICE])
+
+    _LOGGER.debug("CEC setup")
+    config = base_config.get(DOMAIN)
+
+    global CEC_CLIENT
+    CEC_CLIENT = CecClient(hass)
+    exclude = config.get(CONF_EXCLUDE)
+    stop = threading.Event()
 
     def _start_cec(event):
         """Open CEC adapter."""
-        adapters = _CEC.DetectAdapters()
-        if len(adapters) == 0:
-            _LOGGER.error("No CEC adapter found")
-            return
+        # initialise libCEC and enter the main loop
+        if CEC_CLIENT.init_lib_cec():
+            descriptions = load_yaml_config_file(
+                os.path.join(os.path.dirname(__file__), 'services.yaml'))[DOMAIN]
+
+            hass.services.register(DOMAIN, SERVICE_POWER_ON, CEC_CLIENT.power_on, descriptions[SERVICE_POWER_ON],
+                                   SERVICE_POWER_SCHEMA)
+            hass.services.register(DOMAIN, SERVICE_STANDBY, CEC_CLIENT.standby, descriptions[SERVICE_STANDBY],
+                                   SERVICE_POWER_SCHEMA)
+            hass.services.register(DOMAIN, SERVICE_SEND_COMMAND, CEC_CLIENT.tx, descriptions[SERVICE_SEND_COMMAND],
+                                   SERVICE_SEND_COMMAND_SCHEMA)
+            hass.services.register(DOMAIN, SERVICE_VOLUME, CEC_CLIENT.volume, descriptions[SERVICE_VOLUME],
+                                   SERVICE_VOLUME_SCHEMA)
+
+            dev_type = 'switch'
+
+            while True:
+                new_devices = filter(lambda x: CEC_CLIENT.lib_cec.PollDevice(x),
+                                     filter(lambda x: exclude is None or x not in exclude,
+                                            filter(lambda x: x not in DEVICE_PRESENCE or not DEVICE_PRESENCE[x],
+                                                   range(15))))
+                if new_devices:
+                    discovery.load_platform(hass, dev_type, DOMAIN, discovered={ATTR_NEW: new_devices},
+                                            hass_config=base_config)
+                seconds_since_scan = 0
+                while seconds_since_scan < SCAN_INTERVAL:
+                    if stop.is_set():
+                        return
 
-        if _CEC.Open(adapters[0].strComName):
-            hass.services.register(DOMAIN, SERVICE_POWER_ON, _power_on)
-            hass.services.register(DOMAIN, SERVICE_STANDBY, _standby)
-            hass.services.register(DOMAIN, SERVICE_SELECT_DEVICE,
-                                   _select_device)
+                    time.sleep(1)
+                    seconds_since_scan += 1
         else:
-            _LOGGER.error("Failed to open adapter")
+            return False
+
+    def _stop_cec(event):
+        stop.set()
 
     hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_cec)
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _stop_cec)
     return True
+
+
+def icon_by_type(cec_type):
+    return ICON_TV if cec_type == 0 \
```

Comment:
Code would be more readable if you use a dict: `ICONS.get(cec_type, ICON_UNKNOWN)`